### DISCUSSION TITLE
On Air should support sub-apps created with NiceGUI's APIRouter

### DISCUSSION THREAD

**Discussion by xinetzone** [zauberzeug/nicegui#2792] [Ideas / Feature Requests]
```python
ui.run_with(
    app,
    mount_path='/', 
    storage_secret='',  
    on_air=True,
    # native=True,
)
```

see : [main.py](https://github.com/xinetzone/DaoField/blob/main/apps/simple/main.py)

ERROR: ![image](https://github.com/zauberzeug/nicegui/assets/30319120/8d4715a6-665c-4c13-bb63-46ec6fdfd426)

Work: 
- ![image](https://github.com/zauberzeug/nicegui/assets/30319120/9bff092c-1b8a-4fb3-94d9-08ca5062f805)

- ![image](https://github.com/zauberzeug/nicegui/assets/30319120/90bad723-e84c-4816-96ed-bbcc702a34e6)

**Comment by rodja** [comment]
It's working fine for me with this example:

```py
from fastapi import FastAPI
from nicegui import ui

app = FastAPI()

@app.get('/api')
def read_root():
    return {'Hello': 'World'}

@ui.page('/')
def show():
    ui.label('Hello, FastAPI!')
    ui.button('click me', on_click=lambda: ui.notify('clicked'))
    ui.link('subpage', '/subpage')

@ui.page('/subpage')
def subpage():
    ui.label('Hello, subpage!')

ui.run_with(app,on_air=True)
```

Did you enter the url https://asia.on-air.io/test by yourself? Or how did you end up with that? There should always be a path like `/devices/<id>` which identifies your instance.

**Reply by xinetzone** [reply]
```python
# main.py
from fastapi import FastAPI
from nicegui import ui, app as gui
from routers import test

gui.include_router(test.router)

app = FastAPI()

@app.get('/api')
def read_root():
    return {'Hello': 'World'}

@ui.page('/')
def show():
    ui.label('Hello, FastAPI!')
    ui.button('click me', on_click=lambda: ui.notify('clicked'))
    ui.link('subpage', '/subpage')
    ui.link("test", "/test")

@ui.page('/subpage')
def subpage():
    ui.label('Hello, subpage!')

ui.run_with(app, on_air=True)
```

**Reply by xinetzone** [reply]
![image](https://github.com/zauberzeug/nicegui/assets/30319120/5f3ae6cf-bc9f-47c2-b7d1-72bd4ff9bd40)

**Reply by xinetzone** [reply]
```python
# test.py
"""NiceGUI 文档"""
from pathlib import Path
from nicegui import ui, APIRouter

ROOT = Path(__file__).parent
router = APIRouter(prefix='/test')

@router.page('/')
def page():
    ui.label("测试")
```

![image](https://github.com/zauberzeug/nicegui/assets/30319120/4cc7f32a-8ef6-4107-93f3-3459b4f46479)

**Reply by xinetzone** [reply]
![image](https://github.com/zauberzeug/nicegui/assets/30319120/680b7e11-93c7-4472-85f7-a632e385cf80)

**Reply by xinetzone** [reply]
This is an automatic redirect to `https://asia.on-air.io/test`.

**Comment by rodja** [comment]
Oh wow. I think we totally miss On Air support for APIRouter. Here a minimal reproduction:

```py
from nicegui import APIRouter, app, ui

router = APIRouter(prefix='/router')
app.include_router(router)
ui.link('go to router', '/router')
ui.link('go to subpage', '/subpage')

@router.page('/')
def router_page():
    ui.label('Hello from APIRouter!')

@ui.page('/subpage')
def subpage():
    ui.label('Hello from subpage!')

ui.run(on_air=True)
```

Opening subpage works, but via APIRouter not. This is correctly marked as a feature request. I'll just update the title accordingly.

**Comment by rodja** [comment]
The issue seems to be related to https://github.com/Kludex/starlette/issues/869 and https://github.com/Kludex/starlette/discussions/2514. When accessing `/router/` it works, but `/router` results in a 307 redirect to the **absolute** url http://localhost/router/. I still can not understand why Starlette is doing such a thing. But be that as it may: in NiceGUI we can improve upon that. See pull request #5113.
